### DISCUSSION TITLE
Question about the render() method

### DISCUSSION THREAD

**Discussion by NDR008** [trackmania-rl/tmrl#46] [Q&A]
```py
    self.renderingThread = Thread(target=self._renderingThread, args=(), kwargs={}, daemon=True) # instantiated with the gym env

    def _renderingThread(self):
        while True:
            self.render()

    def render(self):
        cv2.imshow('Render Display', self.server.pic)
        if cv2.waitKey(1) & 0xFF == ord('q'):
            return
        return True
```

I'm a bit confused why the above works in this below:
```py
obs, rew, terminated, truncated, info = env.step(env.action_space.sample())
while not (terminated or truncated):
    obs, rew, terminated, truncated, info = env.step(env.action_space.sample())
    env.render() # why is this like making a render appear?
    print(f"rew:{rew}")
    ```

I'm confused why without env.render() no render is show (I know it sounds moronic). I mean the rendering thread is instantiated at the enviornment creation, and the renderingThread() has an infinite self.render() loop, so... 
Brain fade on my end somewhere?

**Comment by NDR008** [comment]
Now I remember where I got this from...
https://github.com/trackmania-rl/tmrl/blob/7fc88de44691077e26e7204f73ced1f6b00f9385/readme/tuto_library.md

```py
    def __init__(self):
        self.rc_drone = None
        self.target = np.array([0.0, 0.0], dtype=np.float32)
        self.initialized = False
        self.blank_image = np.ones((500, 500, 3), dtype=np.uint8) * 255
        self.rendering_thread = Thread(target=self._rendering_thread, args=(), kwargs={}, daemon=True)

    def _rendering_thread(self):
        from time import sleep
        while True:
            sleep(0.1)
            self.render()
            
   def render(self):
        image = self.blank_image.copy()
        pos_x, pos_y = self.rc_drone.get_observation()
        image = cv2.circle(img=image,
                           center=(int(pos_x * 200) + 250, int(pos_y * 200) + 250),
                           radius=10,
                           color=(255, 0, 0),
                           thickness=1)
        image = cv2.circle(img=image,
                           center=(int(self.target[0] * 200) + 250, int(self.target[1] * 200) + 250),
                           radius=5,
                           color=(0, 0, 255),
                           thickness=-1)
        cv2.imshow("Dummy RC drone", image)
        if cv2.waitKey(1) & 0xFF == ord('q'):
            return
            
            ```
            
I do not understand... if the thread is always running, would it not always display even if env.render() is NOT called?
On the otherhand, is env.render() making us of the thread?

**Comment by yannbouteiller** [comment]
You can ignore render() entirely, not implement it and not use it.

It is a Gymnasium method for environments that don't naturally have a display, for example the dummy drone environment of the rtgym tutorial. This method is entirely optional, it is in charge of displaying a window with whatever you want to display in it.

**Reply by NDR008** [reply]
Hmmm need to think. After some time I understood the self stupidity in processing those code snippets.
But.
I do wish to implement this with threads, because I might use a displayless emulater later, to reduce resources.
